lib/Transforms/MemoryBanking.cpp

```diff
@@ -324,43 +320,118 @@ void updateFuncOpArgAttrs(func::FuncOp funcOp, unsigned argIndex,
-                                  std::optional<int> bankingDimensionOpt) {
+unsigned getCurrBankingInfo(BankingConfigAttributes bankingConfigAttrs,
+                            StringRef attrName) {
+  if (attrName.str() == bankingFactorsStr) {
+    if (auto arrayAttr = dyn_cast<ArrayAttr>(bankingConfigAttrs.factors)) {
+      assert(!arrayAttr.empty() &&
+             "BankingConfig factors ArrayAttr should not be empty");
+      auto intAttr = dyn_cast<IntegerAttr>(arrayAttr.getValue().front());
+      assert(intAttr && "BankingConfig factors must be integers");
+      return intAttr.getInt();
+    }
+    auto intAttr = dyn_cast<IntegerAttr>(bankingConfigAttrs.factors);
+    assert(intAttr && "BankingConfig factor must be an integer");
+    return intAttr.getInt();
+  }
+
+  assert(attrName.str() == bankingDimensionsStr &&
+         "BankingConfig only contains 'factors' and 'dimensions' attributes");
+  if (auto arrayAttr = dyn_cast<ArrayAttr>(bankingConfigAttrs.dimensions)) {
+    assert(!arrayAttr.empty() &&
+           "BankingConfig dimensions ArrayAttr should not be empty");
+    auto intAttr = dyn_cast<IntegerAttr>(arrayAttr.getValue().front());
+    assert(intAttr && "BankingConfig dimensions must be integers");
+    return intAttr.getInt();
+  }
+  auto intAttr = dyn_cast<IntegerAttr>(bankingConfigAttrs.dimensions);
+  assert(intAttr && "BankingConfig dimension must be an integer");
+  return intAttr.getInt();
+}
+
+Attribute getRemainingBankingInfo(MLIRContext *context,
+                                  BankingConfigAttributes bankingConfigAttrs,
+                                  StringRef attrName) {
+  if (attrName.str() == bankingFactorsStr) {
+    if (auto arrayAttr = dyn_cast<ArrayAttr>(bankingConfigAttrs.factors)) {
+      assert(!arrayAttr.empty() &&
+             "BankingConfig factors ArrayAttr should not be empty");
+      return arrayAttr.size() > 1
+                 ? ArrayAttr::get(context, arrayAttr.getValue().take_back(
+                                               arrayAttr.size() - 1))
+                 : nullptr;
+    }
+    auto intAttr = dyn_cast<IntegerAttr>(bankingConfigAttrs.factors);
+    assert(intAttr && "BankingConfig factor must be an integer");
+    return nullptr;
+  }
+
+  assert(attrName.str() == bankingDimensionsStr &&
+         "BankingConfig only contains 'factors' and 'dimensions' attributes");
+  if (auto arrayAttr = dyn_cast<ArrayAttr>(bankingConfigAttrs.dimensions)) {
+    assert(!arrayAttr.empty() &&
+           "BankingConfig dimensions ArrayAttr should not be empty");
+    return arrayAttr.size() > 1
+               ? ArrayAttr::get(context, arrayAttr.getValue().take_back(
+                                             arrayAttr.size() - 1))
+               : nullptr;
+  }
+  auto intAttr = dyn_cast<IntegerAttr>(bankingConfigAttrs.dimensions);
+  assert(intAttr && "BankingConfig dimension must be an integer");
+  return nullptr;
+}
```

Comment:
Nit this looks almost exactly the same; can we have a helper function here to handle this? Same with `getCurrBankingInfo`.

Reply:
you are right, I'll try to refactor